deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/config/ZKMysqlDataSourceFactory.java

```diff
@@ -84,9 +99,49 @@ public BasicDataSource build() throws Exception {
             host = reader.getHost(replicaSet);
             port = reader.getPort(replicaSet);
         }
-        KnoxDBKeyReader.init((role));
-        String userName = KnoxDBKeyReader.getUserName();
-        String password = KnoxDBKeyReader.getPassword();
-        return DatabaseUtil.createMysqlDataSource(host, port, userName, password, pool);
+
+        if (this.useMTLS) {
+            String userName = getUserNameFromSpiffeId();
+            String password = "password";
+            String trustPassword = this.trustPasswd;
+            String clientPassword = this.clientPasswd;
+            Map<String, String> proxyConnectionProps = ImmutableMap.<String, String>builder()
+                // ssl properties
+                .put("sslMode", "VERIFY_CA" )
+                .put("trustCertificateKeyStoreUrl", "file:/var/lib/normandie/fuse/jkstrust/generic" )
+                .put("trustCertificateKeyStoreType", "JKS" )
+                .put("trustCertificateKeyStorePassword",  trustPassword)
+                .put("clientCertificateKeyStoreUrl", "file:/var/lib/normandie/fuse/jks/generic" )
+                .put("clientCertificateKeyStoreType", "JKS" )
+                .put("clientCertificateKeyStorePassword",  clientPassword)
+                .build();
+            host = this.replicaSet;
+            // we don't need the replica number in the host; 
+            // if in the configuration we input the number in the replica, we have to remove it.
+            if (host.length() > 3) {
+                String replicaSetNumber = replicaSet.substring(replicaSet.length() - 3);
+                if (StringUtils.isNumeric(replicaSetNumber)) {
+                    host = replicaSet.substring(0, replicaSet.length() - 3);
+                } 
+            }
+            host += ".proxysql.pinadmin.com"; 
+            return DatabaseUtil.createMysqlDataSource(host, port, userName, password, pool, proxyConnectionProps);
+        } else {
+            KnoxDBKeyReader.init((role));
+            String userName = KnoxDBKeyReader.getUserName();
+            String password = KnoxDBKeyReader.getPassword();
+            return DatabaseUtil.createMysqlDataSource(host, port, userName, password, pool);
+        }
+    }
+
+    /**
+     * For mysql 8 and mtls migration we are generating usernames from spiffe's md5 hash
+     *
+     */
+    public static String getUserNameFromSpiffeId() {
```

Comment:
Just realized this function is Pinterest specific too, not sure if we want to do anything about that.

Reply:
> Just realized this function is Pinterest specific too, not sure if we want to do anything about that.

I think this looks good and we can keep what it is. ^^